File: Coding/computer/数据结构/堆/ArrayHeap.py

class ArrayHeap(object):
    """基于列表（数组）实现的最大堆和最小堆数据结构"""
    def __init__(self, heapType = 0):

        """initial"""
        self.headType = heapType
        self.__sentry = -1
        self.__heap = [-1]
        self.__startIndex = 1

    def add(self, val):
        self.__heap.append(val)

        current = self.tailIndex
        parent = current // 2

        while not self.compareTo(self.__heap[parent], self.__heap[current]):
            temp = self.__heap[parent]
            self.__heap[parent] = self.__heap[current]
            self.__heap[current] = temp

            current = parent
            parent = parent // 2

    def delete(self):
        if self.isEmpty:
            raise Exception("Empty Exception")

        root = self.__heap[self.__startIndex]

        self.__heap[self.__startIndex] = self.__heap[self.tailIndex]
        self.__heap.pop()

        currentIndex = self.__startIndex

        while currentIndex <= self.tailIndex:
            leftIndex = currentIndex * 2
            rightIndex = currentIndex * 2 + 1

            if leftIndex > self.tailIndex:
                break
            left = self.__heap[leftIndex]

            if rightIndex > self.tailIndex:
                if self.compareTo(left, self.__heap[currentIndex]):
                    self.__heap[currentIndex], self.__heap[leftIndex] = self.__heap[leftIndex], self.__heap[currentIndex]
                    currentIndex = leftIndex
                else:
                    break
            else:
                right = self.__heap[rightIndex]

                extraIndex = leftIndex
                if not self.compareTo(left, right):
                    extraIndex = rightIndex

                if self.compareTo(self.__heap[extraIndex], self.__heap[currentIndex]):
                    self.__heap[currentIndex], self.__heap[extraIndex] = self.__heap[extraIndex], self.__heap[
                        currentIndex]
                    currentIndex = extraIndex
                else:
                    break

        return root

    def compareTo(self, a, b):
        if self.headType == 0:
            return a < b
        else:
            return b < a

    @property
    def tailIndex(self):
        return len(self.__heap) - 1

    @property
    def count(self):
        return len(self.__heap) - 1

    @property
    def isEmpty(self):
        return len(self.__heap) <= 1

    @property
    def sentry(self):
        return self.__sentry

    @sentry.setter
    def sentry(self, val):
        self.__sentry = val
        self.__heap[0] = val

    def __str__(self):
        return str(self.__heap[1:])


def main():
    myheap = ArrayHeap(1)
    myheap.sentry = 99
    myheap.add(3)
    myheap.add(1)
    myheap.add(8)
    myheap.add(2)
    myheap.add(6)
    myheap.add(5)
    myheap.add(7)
    myheap.add(9)
    myheap.add(0)
    myheap.add(4)
    print(myheap)

    size = myheap.count

    for i in range(size):
        ret = myheap.delete()
        print(ret)

if __name__ == "__main__":
    main()

File: Coding/computer/数据结构/堆/ArrayHeap.cs
/*
最大堆和最小堆的数组实现

 ArrayHeap<int> heap = new ArrayHeap<int>(HeapType.Max);
            heap.Sentry = Int32.MaxValue;
            heap.Add(2);
            heap.Add(1);
            heap.Add(3);
            heap.Add(9);
            heap.Add(8);
            heap.Add(5);
            heap.Add(6);
            heap.Add(4);
            heap.Add(7);

            var count = heap.Count;

            for (int i = 0; i < count; i++)
            {
                var ret = heap.Del();
                Console.WriteLine(ret);
            }
 */

using System;
using System.Collections.Generic;

namespace MyTests
{
    public class ArrayHeap<T> where T: IComparable<T>
    {
        #region Constructors

        /// <summary>
        /// 默认为最小堆
        /// </summary>
        /// <param name="heapType"></param>
        public ArrayHeap(HeapType heapType = HeapType.Min)
        {
            _heapType = heapType;
            _heaps.Add(default(T)); // sentry
        }

        #endregion Constructors

        #region  Fields

        private List<T> _heaps = new List<T>(); // 堆列表


        private int  _startIndex = 1; // 起始位置

        #endregion Fields

        #region Properties

        #region FullProperty HeapType

        private readonly HeapType _heapType;

        public HeapType HeapType
        {
            get { return _heapType; }
        }

        #endregion	FullProperty HeapType

        #region FullProperty Sentry 哨兵

        private T _sentry;

        public T Sentry
        {
            get { return _sentry; }
            set { _sentry = value; }
        }

        #endregion	FullProperty Sentry

        public int Count => _heaps.Count - 1;  // 堆的数量

        public int TailIndex => Count; // 尾巴索引

        public bool IsEmpty => _heaps.Count == 1; // 是否是空的

        #endregion Properties

        #region Methods

        /// <summary>
        /// 比较函数
        /// </summary>
        /// <param name="a"></param>
        /// <param name="b"></param>
        /// <returns></returns>
        public int Compare(T a, T b)
        {
            int flag = 0;
            switch (HeapType)
            {
                case HeapType.Min:
                    flag = a.CompareTo(b);
                    break;
                case HeapType.Max:
                    flag = b.CompareTo(a);
                    break;
                default:
                    break;
            }

            return flag;
        }

        /// <summary>
        /// 添加一个元素
        /// </summary>
        /// <param name="val"></param>

        public void Add(T val)
        {
            _heaps.Add(val); // 首先加到堆列表的末尾
            // 随后调整位置

            int current = TailIndex;
            int parent = current / 2;

            while (Compare(_heaps[parent], _heaps[current]) > 0)
            {
                var temp = _heaps[parent];
                _heaps[parent] = _heaps[current];
                _heaps[current] = temp;

                current = parent;
                parent /= 2;
            }
        }

        public T Del()
        {
            if (IsEmpty)
            {
                throw new Exception("堆已空");
            }

            var root = _heaps[_startIndex];

            _heaps[_startIndex] = _heaps[TailIndex];

            _heaps.RemoveAt(TailIndex); // 删除尾元素

            // 调整结构

            int currentIndex = _startIndex;

            while (currentIndex <= TailIndex)
            {
                int leftIndex = currentIndex * 2;
                int rightIndex = currentIndex * 2 + 1;

                if (leftIndex > TailIndex)
                {
                    break;
                }

                T left = _heaps[leftIndex];

                if (rightIndex > TailIndex)
                {
                    if (Compare(left, _heaps[currentIndex]) < 0)
                    {
                        var temp = _heaps[leftIndex];
                        _heaps[leftIndex] = _heaps[currentIndex];
                        _heaps[currentIndex] = temp;

                        currentIndex = leftIndex;
                    }
                    else
                    {
                        break;
                    }
                }
                else
                {
                    T right = _heaps[rightIndex];

                    var extraIndex = leftIndex;
                    if (Compare(left, right) > 0)
                    {
                        extraIndex = rightIndex;
                    }

                    if (Compare(_heaps[extraIndex], _heaps[currentIndex]) < 0)
                    {
                        var temp = _heaps[extraIndex];
                        _heaps[extraIndex] = _heaps[currentIndex];
                        _heaps[currentIndex] = temp;

                        currentIndex = extraIndex;
                    }
                    else
                    {
                        break;
                    }
                }
            }

            return root;
        }

        private bool CheckSons(int parentIndex)
        {
            int leftSon = parentIndex * 2;
            int rightSon = parentIndex * 2 + 1;

            if (leftSon > TailIndex)
            {
                return false;
            }

            // 右儿子超出索引
            if (rightSon > TailIndex)
            {
                return Compare(_heaps[leftSon], _heaps[parentIndex]) > 0;
            }
            else
            {
                return Compare(_heaps[leftSon], _heaps[parentIndex]) > 0 &&
                       Compare(_heaps[rightSon], _heaps[parentIndex]) > 0;
            }
        }
            

        #endregion Methods
    }

    public enum HeapType
    {
        Min = 0,
        Max
    }
}
